backfill.go

```diff
@@ -0,0 +1,123 @@
+type BackfillRequester interface {
+	// ServersAtEvent is called when trying to determine which server to request from.
+	// It returns a list of servers which can be queried for backfill requests. These servers
+	// will be servers that are in the room already. The entries at the beginning are preferred servers
+	// and will be tried first. An empty list will fail the request.
+	ServersAtEvent(ctx context.Context, roomID, eventID string) []ServerName
+	// Backfill performs a backfill request to the given server.
+	// https://matrix.org/docs/spec/server_server/latest#get-matrix-federation-v1-backfill-roomid
+	Backfill(ctx context.Context, server ServerName, roomID string, fromEventIDs []string, limit int) (*Transaction, error)
+	// StateIDs performs a state IDs request to the given server.
+	// https://matrix.org/docs/spec/server_server/latest#get-matrix-federation-v1-state-ids-roomid
+	StateIDs(ctx context.Context, server ServerName, roomID, eventID string) (*RespStateIDs, error)
+	// EventAuth performs an event auth request to the given server.
+	// https://matrix.org/docs/spec/server_server/latest#get-matrix-federation-v1-event-auth-roomid-eventid
+	EventAuth(ctx context.Context, server ServerName, roomID, eventID string) (*RespEventAuth, error)
+}
+
+// RequestBackfill implements the server logic for making backfill requests to other servers.
+// This handles server selection, fetching up to the request limit and verifying the received events.
+// Event validation also includes authorisation checks, which may require additional state to be fetched.
+//
+// The returned events are safe to be inserted into a database for later retrieval. It's possible for the
+// number of returned events to be less than the limit, even if there exists more events. It's also possible
+// for the number of returned events to be greater than the limit, if fromEventIDs > 1 and we need to ask
+// multiple servers. We don't drop events greater than the limit because we've already done all the work to
+// verify them, so it's up to the caller to decide what to do with them.
+//
+// TODO: We should be able to make some guarantees for the caller about the returned events position in the DAG,
+// but to verify it we need to know the prev_events of fromEventIDs.
+//
+// TODO: When does it make sense to return errors?
+func RequestBackfill(ctx context.Context, b BackfillRequester, keyRing JSONVerifier,
+	roomID string, ver RoomVersion, fromEventIDs []string, limit int) ([]HeaderedEvent, error) {
+
+	if len(fromEventIDs) == 0 {
+		return nil, nil
+	}
+	haveEventIDs := make(map[string]bool)
+	var result []HeaderedEvent
+	// pick a server to backfill from
+	// TODO: use other event IDs and make a set out of all the returned servers?
+	servers := b.ServersAtEvent(ctx, roomID, fromEventIDs[0])
+	// loop each server asking it for `limit` events. Worst case, we ask every server for `limit`
+	// events before giving up. Best case, we just ask one.
+	for _, s := range servers {
```

Comment:
This is something we're generally quite bad at throughout, but it just stood out to me here, that wherever we have HTTP requests in loops where we `continue` on error, we really should check that the initial context hasn't expired at the beginning of each iteration to save us from unnecessarily calling `Backfill` a bunch more times with an expired context. It's basically a no-op if we do but they are no-ops that are noisy in the logs.

Reply:
Good point, done!